source/MaterialXGenMdl/Nodes/ClosureCompoundNodeMdl.cpp

```diff
@@ -30,12 +30,75 @@ void ClosureCompoundNodeMdl::emitFunctionDefinition(const ShaderNode& node, GenC
     {
         const ShaderGenerator& shadergen = context.getShaderGenerator();
 
-        const bool isMaterialExpr = (_rootGraph->hasClassification(ShaderNode::Classification::CLOSURE) ||
-                                     _rootGraph->hasClassification(ShaderNode::Classification::SHADER));
-
         // Emit functions for all child nodes
         shadergen.emitFunctionDefinitions(*_rootGraph, context, stage);
 
+        // split all fields into separate functions
+        std::unordered_set<const ShaderNode*> knownUpsteams;
+        if (!_returnStruct.empty() && _unrollReturnStructMembers)
+        {
+            // make sure the upstream definitions are known
+            for (size_t i = 0, n = _rootGraph->numOutputSockets(); i < n; ++i)
```

Comment:
Nitpick: To harmonize coding convention the loop could be written as:

`for (const ShaderGraphOutputSocket* output : _rootGraph->getOutputSockets())`